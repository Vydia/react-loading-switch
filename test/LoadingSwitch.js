// @flow
/* eslint-disable react/no-multi-comp */
const { forEach } = require('lodash')
const LoadingSwitch = require('../lib/LoadingSwitch')
const isPending = require('../lib/utils/isPending')
const renderer = require('react-test-renderer')
const React = require('react')

const renderError = ({ message }) => `Error: ${message}`
renderError.displayName = 'renderError'

const renderLoading = () => 'Loading'
renderLoading.displayName = 'renderLoading'

const children = () => 'Children'
children.displayName = 'children'

const loading = true
const error = new Error('`error`: Test message')
const errorWhenMissing = () => new Error('`errorWhenMissing`: Test message')

const propsMissingData = {
  errorWhenMissing,
  loading: false,
  renderError,
  require: undefined,
}

forEach([
  {
    description: 'renders `children()` when `require` is truthy',
    props: {
      children,
      require: 'whatever',
    },
  },
  {
    description: 'renders `renderLoading()` when `loading` is true and `require` is undefined',
    props: {
      loading,
      renderLoading,
      require: undefined,
    },
  },
  {
    description: 'renders `renderLoading()` when `loading` is true and `require` is false',
    props: {
      loading,
      renderLoading,
      require: false,
    },
  },
  {
    description: 'renders `renderLoading()` when `loading` is true and `require` is 0',
    props: {
      loading,
      renderLoading,
      require: 0,
    },
  },
  {
    description: 'renders `renderError(error)` when `error` is provided',
    props: {
      error,
      renderError,
    },
  },
  {
    description: 'renders `renderError(errorWhenMissing())` when `loading` is false but `require` is still falsey',
    props: {
      ...propsMissingData,
    },
  },
  {
    description: 'renders `renderError(error))` when both `error` and `errorWhenMissing` are provided',
    props: {
      ...propsMissingData,
      errorWhenMissing,
      error,
    },
  },
], ({ description, props: { children, ...props } }) => {
  it(description, () => {
    const element = React.createElement(LoadingSwitch, props, children)
    expect(renderer.create(element).toJSON()).toMatchSnapshot()
  })
})

// TODO: Move to separate file.
describe('isPending', () => {
  forEach([
    {
      param: undefined,
      expected: true,
    },
    {
      param: null,
      expected: true,
    },
    {
      param: false,
      expected: true,
    },
    {
      param: 0,
      expected: true,
    },
    {
      param: '',
      expected: true,
    },
    {
      param: [],
      expected: false,
    },
    {
      param: {},
      expected: false,
    },
    {
      param: 'whatever',
      expected: false,
    },
    {
      param: { foo: undefined },
      expected: false,
    },
    {
      param: { foo: null },
      expected: false,
    },
  ], ({ param, expected }) => {
    describe(`when given ${JSON.stringify(param)}`, () => {
      it(`returns ${expected}`, () => {
        expect(isPending(param)).toEqual(expected)
      })
    })
  })
})
