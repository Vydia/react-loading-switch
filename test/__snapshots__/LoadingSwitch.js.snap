// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders \`children()\` when \`require\` is truthy 1`] = `"Children"`;

exports[`renders \`renderError(error))\` when both \`error\` and \`errorWhenMissing\` are provided 1`] = `"Error: \`error\`: Test message"`;

exports[`renders \`renderError(error)\` when \`error\` is provided 1`] = `"Error: \`error\`: Test message"`;

exports[`renders \`renderError(errorWhenMissing())\` when \`loading\` is false but \`require\` is still falsey 1`] = `"Error: \`errorWhenMissing\`: Test message"`;

exports[`renders \`renderLoading()\` when \`loading\` is true and \`require\` is 0 1`] = `"Loading"`;

exports[`renders \`renderLoading()\` when \`loading\` is true and \`require\` is false 1`] = `"Loading"`;

exports[`renders \`renderLoading()\` when \`loading\` is true and \`require\` is undefined 1`] = `"Loading"`;
